File: packages/python/sharing/starfish_sharing/directory.py
"""Member directory helpers (Python mirror of TS ``sharing/directory.ts``).

One entry per ``kind: "member"`` cap the owner has issued for a given
collection. The doc lives at ``<collectionPath>/_members`` and is
**owner-only**: non-admin member caps are rejected at well-formedness time
when they would reach ``<col>/_members``.

Device directory helpers live in ``starfish_identities.directory``.
"""

from __future__ import annotations

import time
from typing import TYPE_CHECKING, Any, Optional, TypedDict

from starfish_protocol.cap import CapCert, recipient_kem
from starfish_sdk.types import ConflictError, StarfishHttpError

if TYPE_CHECKING:
    from starfish_sdk.client import StarfishClient


class DirectoryEntry(TypedDict, total=False):
    """A single entry in a device or member directory document.

    ``cert`` carries the full signed cap-cert so a ``_members`` roster can double
    as a cap-distribution list for plaintext (cap-only) collections: the owner
    publishes each member's signed cap here and members fetch their own with
    ``fetch_my_member_cap``. The
    other fields are a redundant projection kept for back-compat and cheap
    ``exp``/``subUserId`` filtering. Safe to publish — a cap is usable only by the
    holder of its subject (``sub``) private key (the server verifies each request
    against ``cert.sub``), so a readable roster never lets one member act as another.
    """

    nonce: str
    sub: str
    subKem: str
    subKemAlg: str
    subUserId: str
    scope: dict[str, Any]
    nbf: int
    exp: int
    label: str
    addedBy: str
    addedAt: int
    cert: dict[str, Any]


class Directory(TypedDict):
    """Directory document stored at the conventional path."""

    v: int
    entries: list[DirectoryEntry]


def members_path_for(collection_path: str) -> str:
    """Returns the storage path for a collection's members directory."""
    return f"{collection_path}/_members"


_MAX_RETRIES = 3


async def _pull_directory(
    client: StarfishClient, path: str
) -> tuple[Directory, Optional[str]]:
    try:
        result = await client.pull(f"/pull/{path}")
    except StarfishHttpError as exc:
        if exc.status == 404:
            return ({"v": 1, "entries": []}, None)
        raise
    return (result.data, result.hash)  # type: ignore[union-attr,return-value]


def _entry_from_cert(
    cert: CapCert | dict[str, Any],
    *,
    label: Optional[str] = None,
    added_by: Optional[str] = None,
) -> DirectoryEntry:
    cert_d: dict[str, Any] = dict(cert)  # type: ignore[arg-type]
    # The ``_members`` directory records single-subject grants. ``recipient_kem``
    # resolves the recipient's KEM pubkey + suite (the dedicated ``subKem`` for
    # ed25519/mixed pairs, or the signing ``sub`` for same-suite secp256k1); it
    # raises for an audience cap, distributed as a public link, not a roster entry.
    if cert_d.get("sub") is None:
        raise ValueError(
            "cannot publish a subject-less cap (e.g. audience) to the member directory"
        )
    kem_pub_hex, kem_alg = recipient_kem(cert_d)
    entry: DirectoryEntry = {
        "nonce": cert_d["nonce"],
        "sub": cert_d["sub"],
        "subKem": kem_pub_hex,
        "scope": cert_d["scope"],
        "nbf": cert_d["nbf"],
        "exp": cert_d["exp"],
        "addedAt": int(time.time()),
        "cert": cert_d,
    }
    if kem_alg != "ed25519":
        entry["subKemAlg"] = kem_alg
    if "subUserId" in cert_d:
        entry["subUserId"] = cert_d["subUserId"]
    if label is not None:
        entry["label"] = label
    if added_by is not None:
        entry["addedBy"] = added_by
    return entry


async def _upsert_entry(
    client: StarfishClient,
    path: str,
    cert: CapCert | dict[str, Any],
    *,
    label: Optional[str] = None,
    added_by: Optional[str] = None,
) -> None:
    cert_nonce = dict(cert)["nonce"]  # type: ignore[arg-type]
    last_err: Optional[Exception] = None
    for attempt in range(_MAX_RETRIES):
        directory, base_hash = await _pull_directory(client, path)
        filtered = [e for e in directory["entries"] if e["nonce"] != cert_nonce]
        new_entry = _entry_from_cert(cert, label=label, added_by=added_by)
        next_doc: Directory = {"v": 1, "entries": [*filtered, new_entry]}
        try:
            await client.push(f"/push/{path}", dict(next_doc), base_hash)
            return
        except ConflictError as exc:
            last_err = exc
            if attempt < _MAX_RETRIES - 1:
                continue
            raise
    if last_err is not None:
        raise last_err


async def _remove_entry_by_nonce(
    client: StarfishClient, path: str, nonce: str
) -> bool:
    last_err: Optional[Exception] = None
    for attempt in range(_MAX_RETRIES):
        directory, base_hash = await _pull_directory(client, path)
        if not any(e["nonce"] == nonce for e in directory["entries"]):
            return False
        next_doc: Directory = {
            "v": 1,
            "entries": [e for e in directory["entries"] if e["nonce"] != nonce],
        }
        try:
            await client.push(f"/push/{path}", dict(next_doc), base_hash)
            return True
        except ConflictError as exc:
            last_err = exc
            if attempt < _MAX_RETRIES - 1:
                continue
            raise
    if last_err is not None:
        raise last_err
    return False


class ListDirectoryOpts(TypedDict, total=False):
    """Optional filters applied at list time."""

    include_expired: bool
    revoked_nonces: frozenset[str] | set[str]


async def _list_entries(
    client: StarfishClient,
    path: str,
    opts: Optional[ListDirectoryOpts] = None,
) -> list[DirectoryEntry]:
    opts = opts or {}
    include_expired = opts.get("include_expired", False)
    revoked = opts.get("revoked_nonces") or set()
    directory, _ = await _pull_directory(client, path)
    now = int(time.time())
    out: list[DirectoryEntry] = []
    for e in directory["entries"]:
        if not include_expired and e.get("exp", 0) < now:
            continue
        if e["nonce"] in revoked:
            continue
        out.append(e)
    return out


async def add_member_entry(
    client: StarfishClient,
    collection_path: str,
    cert: CapCert | dict[str, Any],
    *,
    label: Optional[str] = None,
    added_by: Optional[str] = None,
) -> None:
    """Append (or overwrite by nonce) a member cap-cert entry in the directory
    at ``<collectionPath>/_members``. Pull-merge-push with baseHash retry.

    Raises ``ValueError`` when ``cert["kind"] != "member"``.
    """
    cert_d = dict(cert)  # type: ignore[arg-type]
    if cert_d.get("kind") != "member":
        raise ValueError(
            f"add_member_entry: expected kind='member', got kind={cert_d.get('kind')!r}"
        )
    await _upsert_entry(
        client,
        members_path_for(collection_path),
        cert,
        label=label,
        added_by=added_by,
    )


async def list_members(
    client: StarfishClient,
    collection_path: str,
    opts: Optional[ListDirectoryOpts] = None,
) -> list[DirectoryEntry]:
    """List members recorded for ``collection_path``."""
    return await _list_entries(client, members_path_for(collection_path), opts)


async def remove_member_entry(
    client: StarfishClient, collection_path: str, nonce: str
) -> bool:
    """Remove a member entry by nonce."""
    return await _remove_entry_by_nonce(
        client, members_path_for(collection_path), nonce
    )


# ── Published caps ───────────────────────────────────────────────────────────
#
# In the plaintext, cap-only sharing mode there is no keyring. Instead the owner
# PUBLISHES each member's full signed cap into the single ``<col>/_members`` list
# (configure that collection read-open + owner-only write). A member fetches its
# own cap from there — no out-of-band forwarding — and presents it for content.
# Safe even when world-readable: a cap is bound to its subject key (the server
# verifies each request against ``cert.sub``), so reading another member's cap is
# useless without their private key.


async def publish_member_cap(
    client: StarfishClient,
    collection_path: str,
    cert: CapCert | dict[str, Any],
    *,
    label: Optional[str] = None,
    added_by: Optional[str] = None,
) -> None:
    """Publish a member's full signed cap into ``<collection_path>/_members``.

    Intention-revealing alias for ``add_member_entry`` (which stores the full
    ``cert``). Owner-only by collection write-roles. Raises on a non-member cap.
    """
    await add_member_entry(
        client, collection_path, cert, label=label, added_by=added_by
    )


async def fetch_member_caps(
    client: StarfishClient,
    collection_path: str,
    opts: Optional[ListDirectoryOpts] = None,
) -> list[dict[str, Any]]:
    """Fetch every published member cap from ``<collection_path>/_members``.

    Returns the usable signed caps (skipping legacy entries written before caps
    were stored). Honors ``ListDirectoryOpts`` (``include_expired``/``revoked_nonces``).
    """
    entries = await _list_entries(
        client, members_path_for(collection_path), opts
    )
    return [e["cert"] for e in entries if e.get("cert") is not None]


async def fetch_my_member_cap(
    client: StarfishClient,
    collection_path: str,
    my_ed_pub_hex: str,
    opts: Optional[ListDirectoryOpts] = None,
) -> Optional[dict[str, Any]]:
    """Fetch the caller's own published cap — the one whose ``sub`` equals their
    Ed25519 pubkey (hex). Returns ``None`` when none is published for them.
    """
    caps = await fetch_member_caps(client, collection_path, opts)
    for cap in caps:
        if cap.get("sub") == my_ed_pub_hex:
            return cap
    return None


async def unpublish_member_cap(
    client: StarfishClient, collection_path: str, nonce: str
) -> bool:
    """Remove a published cap by nonce (e.g. on eviction). Alias for
    ``remove_member_entry``. Returns ``False`` when the nonce was not present.
    """
    return await remove_member_entry(client, collection_path, nonce)


__all__ = [
    "Directory",
    "DirectoryEntry",
    "ListDirectoryOpts",
    "add_member_entry",
    "list_members",
    "members_path_for",
    "remove_member_entry",
    "publish_member_cap",
    "fetch_member_caps",
    "fetch_my_member_cap",
    "unpublish_member_cap",
]

File: packages/ts/sharing/src/directory.ts
/**
 * Member directory helpers — owner-side audit/UI metadata for
 * `<collectionPath>/_members`.
 *
 * One entry per `kind: "member"` cap the owner has issued for a given
 * collection. The doc is **owner-only**: non-admin member caps are rejected at
 * well-formedness time when they would reach `<col>/_members` (see the
 * `member-members-not-denied` rule).
 *
 * The directory is **not** an authority source. Auth flows entirely through the
 * cap-cert presented in `Authorization` headers and the
 * `_revocations/{rootUserId}` document.
 *
 * Device directory helpers (`addDeviceEntry`, `listDevices`, …) live in
 * `@drakkar.software/starfish-identities`.
 */

import type { Alg, CapCert } from "@drakkar.software/starfish-protocol"
import { recipientKem } from "@drakkar.software/starfish-protocol"
import type { StarfishClient } from "@drakkar.software/starfish-client"
import { ConflictError, StarfishHttpError } from "@drakkar.software/starfish-client"

// ── Types & paths ────────────────────────────────────────────────────────────

/** A single entry in a device or member directory document. */
export interface DirectoryEntry {
  /** Base64 cap nonce — identifies the cert. */
  nonce: string
  /** Subject signing pubkey, hex (32 B), of suite `subAlg`. */
  sub: string
  /** Subject KEM pubkey, hex (32 B), of suite `subKemAlg` — the keyring recipient key. */
  subKem: string
  /** Subject KEM suite. Absent ⇒ `ed25519` (X25519). */
  subKemAlg?: Alg
  /** `sha256(sub)[0:32]`. Required for member caps; undefined for device caps. */
  subUserId?: string
  /** Cap's `scope` block, mirrored verbatim. */
  scope: CapCert["scope"]
  /** Cap's not-before (unix seconds). */
  nbf: number
  /** Cap's expiry (unix seconds). */
  exp: number
  /** Optional human-readable label, e.g. "Bob (designer)". */
  label?: string
  /** Ed25519 pubkey of the device that wrote this entry, hex (optional). */
  addedBy?: string
  /** Unix seconds when the entry was added. */
  addedAt: number
  /**
   * The full signed cap-cert. Present so a `_members` roster can double as a
   * cap-distribution list for plaintext (cap-only) collections: the owner
   * publishes each member's signed cap here and members fetch their own with
   * {@link fetchMyMemberCap}.
   * The other fields above are a redundant projection kept for back-compat and
   * for cheap `exp`/`subUserId` filtering. Safe to publish — a cap is usable
   * only by the holder of its subject (`sub`) private key (the server verifies
   * each request's signature against `cert.sub`), so a readable roster of caps
   * never lets one member act as another.
   */
  cert?: CapCert
}

/** Directory document stored at the conventional path. */
export interface Directory {
  v: 1
  entries: DirectoryEntry[]
}

/** Directory entry for a `kind: "member"` cap. */
export type MemberEntry = DirectoryEntry

/** Returns the storage path for a collection's members directory. */
export function membersPathFor(collectionPath: string): string {
  return `${collectionPath}/_members`
}

// ── HTTP helpers (intentionally duplicated from identities.directory) ────────

async function pullDirectory(
  client: StarfishClient,
  path: string,
): Promise<{ directory: Directory; hash: string | null }> {
  try {
    const result = await client.pull(`/pull/${path}`)
    return {
      directory: result.data as unknown as Directory,
      hash: result.hash,
    }
  } catch (err) {
    if (err instanceof StarfishHttpError && err.status === 404) {
      return { directory: { v: 1, entries: [] }, hash: null }
    }
    throw err
  }
}

function entryFromCert(
  cert: CapCert,
  label?: string,
  addedBy?: string,
): DirectoryEntry {
  // The `_members` directory records single-subject grants. `recipientKem`
  // resolves the recipient's KEM pubkey + suite (the dedicated `subKem` for
  // ed25519/mixed pairs, or the signing `sub` for same-suite secp256k1); it
  // throws for an audience cap, which is distributed as a public link, not a
  // roster entry.
  if (cert.sub === undefined) {
    throw new Error("cannot publish a subject-less cap (e.g. audience) to the member directory")
  }
  const { kemPubHex, kemAlg } = recipientKem(cert)
  const entry: DirectoryEntry = {
    nonce: cert.nonce,
    sub: cert.sub,
    subKem: kemPubHex,
    scope: cert.scope,
    nbf: cert.nbf,
    exp: cert.exp,
    addedAt: Math.floor(Date.now() / 1000),
    cert,
    ...(kemAlg !== "ed25519" ? { subKemAlg: kemAlg } : {}),
  }
  if (cert.subUserId !== undefined) entry.subUserId = cert.subUserId
  if (label !== undefined) entry.label = label
  if (addedBy !== undefined) entry.addedBy = addedBy
  return entry
}

const MAX_RETRIES = 3

async function upsertEntry(
  client: StarfishClient,
  path: string,
  cert: CapCert,
  opts: { label?: string; addedBy?: string },
): Promise<void> {
  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    const { directory, hash } = await pullDirectory(client, path)
    const filtered = directory.entries.filter((e) => e.nonce !== cert.nonce)
    const next: Directory = {
      v: 1,
      entries: [...filtered, entryFromCert(cert, opts.label, opts.addedBy)],
    }
    try {
      await client.push(
        `/push/${path}`,
        next as unknown as Record<string, unknown>,
        hash,
      )
      return
    } catch (err) {
      if (err instanceof ConflictError && attempt < MAX_RETRIES - 1) {
        continue
      }
      throw err
    }
  }
  throw new Error(`directory.upsertEntry: too many baseHash conflicts at ${path}`)
}

async function removeEntryByNonce(
  client: StarfishClient,
  path: string,
  nonce: string,
): Promise<boolean> {
  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    const { directory, hash } = await pullDirectory(client, path)
    if (!directory.entries.some((e) => e.nonce === nonce)) return false
    const next: Directory = {
      v: 1,
      entries: directory.entries.filter((e) => e.nonce !== nonce),
    }
    try {
      await client.push(
        `/push/${path}`,
        next as unknown as Record<string, unknown>,
        hash,
      )
      return true
    } catch (err) {
      if (err instanceof ConflictError && attempt < MAX_RETRIES - 1) {
        continue
      }
      throw err
    }
  }
  throw new Error(`directory.removeEntryByNonce: too many baseHash conflicts at ${path}`)
}

/** Optional filters applied at list time. */
export interface ListDirectoryOpts {
  /** When true, expired entries are returned too (default: false). */
  includeExpired?: boolean
  /** When set, entries whose nonce appears here are filtered out. */
  revokedNonces?: ReadonlySet<string>
}

async function listEntries(
  client: StarfishClient,
  path: string,
  opts: ListDirectoryOpts,
): Promise<DirectoryEntry[]> {
  const { directory } = await pullDirectory(client, path)
  const now = Math.floor(Date.now() / 1000)
  return directory.entries.filter((e) => {
    if (!opts.includeExpired && e.exp < now) return false
    if (opts.revokedNonces?.has(e.nonce)) return false
    return true
  })
}

// ── Members ──────────────────────────────────────────────────────────────────

/**
 * Append (or overwrite by nonce) a member cap-cert entry in the directory at
 * `<collectionPath>/_members`. Pull-merge-push with baseHash retry.
 *
 * `collectionPath` is the storage prefix used by the collection — for a
 * collection whose data lives at `notes/<doc>`, pass `"notes"`. For a
 * per-owner collection at `users/<owner>/notes/<doc>`, pass
 * `"users/<owner>/notes"`.
 *
 * Throws when `cert.kind !== "member"`. The path is only writable by admin
 * or root-device caps; member caps are denied this path by the
 * `member-members-not-denied` well-formedness rule.
 */
export async function addMemberEntry(
  client: StarfishClient,
  collectionPath: string,
  cert: CapCert,
  opts: { label?: string; addedBy?: string } = {},
): Promise<void> {
  if (cert.kind !== "member") {
    throw new Error(
      `addMemberEntry: expected kind="member", got kind="${cert.kind}"`,
    )
  }
  await upsertEntry(client, membersPathFor(collectionPath), cert, opts)
}

/**
 * List members recorded for `collectionPath`. Owner-only by virtue of cap
 * scope — invoking this with a member cap returns 403 at the server. Returns
 * the empty list when no directory document exists yet.
 */
export async function listMembers(
  client: StarfishClient,
  collectionPath: string,
  opts: ListDirectoryOpts = {},
): Promise<MemberEntry[]> {
  return listEntries(client, membersPathFor(collectionPath), opts)
}

/**
 * Remove a member entry by nonce. Same caveat as device removal: this only
 * updates the directory; the server accepts the cap until it's added to the
 * revocation list and its TTL fires.
 */
export async function removeMemberEntry(
  client: StarfishClient,
  collectionPath: string,
  nonce: string,
): Promise<boolean> {
  return removeEntryByNonce(client, membersPathFor(collectionPath), nonce)
}

// ── Published caps ───────────────────────────────────────────────────────────
//
// In the plaintext, cap-only sharing mode there is no keyring. Instead the owner
// PUBLISHES each member's full signed cap into the single `<col>/_members` list
// (configure that collection read-open + owner-only write). A member fetches its
// own cap from there — no out-of-band forwarding — and presents it for content.
// This is safe even when the list is world-readable: a cap is bound to its
// subject key (the server verifies each request against `cert.sub`), so reading
// another member's cap is useless without their private key.

/**
 * Publish a member's full signed cap into `<collectionPath>/_members`. Thin,
 * intention-revealing alias for {@link addMemberEntry} (which already stores the
 * full `cert`). Owner-only by collection write-roles. Throws on a non-member cap.
 */
export async function publishMemberCap(
  client: StarfishClient,
  collectionPath: string,
  cert: CapCert,
  opts: { label?: string; addedBy?: string } = {},
): Promise<void> {
  await addMemberEntry(client, collectionPath, cert, opts)
}

/**
 * Fetch every published member cap from `<collectionPath>/_members`. Returns the
 * usable signed `CapCert`s (skipping legacy entries written before caps were
 * stored). Honors {@link ListDirectoryOpts} (`includeExpired`, `revokedNonces`).
 */
export async function fetchMemberCaps(
  client: StarfishClient,
  collectionPath: string,
  opts: ListDirectoryOpts = {},
): Promise<CapCert[]> {
  const entries = await listEntries(client, membersPathFor(collectionPath), opts)
  return entries
    .map((e) => e.cert)
    .filter((c): c is CapCert => c !== undefined)
}

/**
 * Fetch the caller's own published cap — the one whose `sub` equals their
 * Ed25519 pubkey (hex). Returns `null` when no usable cap is published for them.
 */
export async function fetchMyMemberCap(
  client: StarfishClient,
  collectionPath: string,
  myEdPubHex: string,
  opts: ListDirectoryOpts = {},
): Promise<CapCert | null> {
  const caps = await fetchMemberCaps(client, collectionPath, opts)
  return caps.find((c) => c.sub === myEdPubHex) ?? null
}

/**
 * Remove a published cap by nonce (e.g. on eviction). Alias for
 * {@link removeMemberEntry}. Returns `false` when the nonce was not present.
 */
export async function unpublishMemberCap(
  client: StarfishClient,
  collectionPath: string,
  nonce: string,
): Promise<boolean> {
  return removeMemberEntry(client, collectionPath, nonce)
}
